streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +52,215 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
     List<StreamTask> allTasks() {
         final List<StreamTask> tasks = super.allTasks();
         tasks.addAll(restoring.values());
+        tasks.addAll(suspended.values());
         return tasks;
     }
 
     @Override
     Set<TaskId> allAssignedTaskIds() {
         final Set<TaskId> taskIds = super.allAssignedTaskIds();
         taskIds.addAll(restoring.keySet());
+        taskIds.addAll(suspended.keySet());
         return taskIds;
     }
 
     @Override
     boolean allTasksRunning() {
-        return super.allTasksRunning() && restoring.isEmpty();
+        return super.allTasksRunning() && restoring.isEmpty() && suspended.isEmpty();
     }
 
-    RuntimeException closeAllRestoringTasks() {
+    @Override
+    void closeTask(final StreamTask task, final boolean clean) {
+        if (suspended.containsKey(task.id())) {
+            task.closeSuspended(clean, false, null);
+        } else {
+            task.close(clean, false);
+        }
+    }
+    
+    Set<TaskId> suspendedTaskIds() {
+        return suspended.keySet();
+    }
+
+    RuntimeException suspend(final Set<TaskId> revokedTasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+
+        log.trace("Suspending running {} {}", taskTypeName, runningTaskIds());
```

Comment:
Are these logs right? It looks like they're saying one thing and doing another. For example, this says it's suspending the running tasks, and lists them, but then actually suspends the _revoked_ tasks. Likewise with the next log.